src/main/java/org/apache/sysds/runtime/matrix/data/LibMatrixReorg.java

```diff
@@ -2243,7 +2271,70 @@ private static void reverseSparse(MatrixBlock in, MatrixBlock out) {
 			if( !a.isEmpty(i) )
 				c.set(m-1-i, a.get(i), true);
 	}
-	
+
+	private static void rollDense(MatrixBlock in, MatrixBlock out, int shift) {
+		final int m = in.rlen;
+		final int n = in.clen;
+
+		//set basic meta data and allocate output
+		out.sparse = false;
+		out.nonZeros = in.nonZeros;
+		out.allocateDenseBlock(false);
+
+		//copy all rows into target positions
+		if( n == 1 ) { //column vector
+			double[] a = in.getDenseBlockValues();
+			double[] c = out.getDenseBlockValues();
+
+			// roll matrix with axis=none
+			shift %= (m != 0 ? m : 1);
+
+			System.arraycopy(a, 0, c, shift, m - shift);
+			System.arraycopy(a, m - shift, c, 0, shift);
+		} else { //general matrix case
+			DenseBlock a = in.getDenseBlock();
+			DenseBlock c = out.getDenseBlock();
+
+			// roll matrix with axis=0
+			shift %= (m != 0 ? m : 1);
+
+			for( int i=0; i< m-shift; i++ ) {
+				System.arraycopy(a.values(i), a.pos(i), c.values(i + shift), c.pos(i + shift), n);
+			}
+
+			for( int i=m-shift; i<m; i++ ) {
+				System.arraycopy(a.values(i), a.pos(i), c.values(i + shift -m), c.pos(i + shift -m), n);
+			}
+		}
+	}
+
+	private static void rollSparse(MatrixBlock in, MatrixBlock out, int shift) {
+		final int m = in.rlen;
+
+		//set basic meta data and allocate output
+		out.sparse = true;
+		out.nonZeros = in.nonZeros;
+		out.allocateSparseRowsBlock(false);
+
+		//copy all rows into target positions
+		SparseBlock a = in.getSparseBlock();
+		SparseBlock c = out.getSparseBlock();
+
+		// roll matrix with axis=0
+		shift %= (m != 0 ? m : 1);
+
+		for( int i=0; i<m-shift; i++ ) {
```

Comment:
you can improve the sparse implementation by processing only non zero elements:
Use something like:

(assuming i is a given row.)
```java
if(a.isEmpty(i))
	continue;
final int apos = sb.pos(i);
final int alen = sb.size(i) + apos;
final int[] aix = sb.indexes(i);
final double[] avals = sb.values(i);
for(int k = apos; k < alen; k++)
    ...
```

Reply:
Modified the code to copy only non-zero elements.

Reply:
Great, it can be further optimized, by minor tweaks such as:

1. Move the inner parts of the for loop to a separate function to enable JIT (the two loops can call the same function with different shift arguments)
2. Move the shift addition out of the inner inner loop but to a temporary variable

Reply:
I appreciate you pointing out the optimization points I hadn’t considered, and I’ve reflected those in the revisions.